Assert mkey as local to the environment and add mkeyvalues to measurement-map. (#257)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

are specific to that instance.

The supply chain entity that is defined as the authority responsible for providing the measurements (i.e. Reference Values or Endorsed Values)
is by default the CoRIM signer.
If different authorities may provide measurement values, the `&(fixme-authorities: -1)` statement can be supplied in the `measurement-map`.
If the `&(fixme-authorities: -1)` statement is used within a condition, at least one of the listed authorities must match the respective ACS entry for the statement to match.

~~~ cddl
{::include cddl/measurement-map.cddl}
~~~

The following describes each member of the `measurement-map`:

* `fixme-authorities"`: The cryptographic identities of individuals or organizations that are
 designated authorities for this measurement. For example, producer of the measurement or a delegated supplier.

* `$measured-element-type-choice`: A map key distinct from `fixme-authorities` that names a measured element. See {{sec-comid-mkey}}.

* `measurement-values-map`: A map value that associates measurement values with the named measured element within the environment. Described in {{sec-comid-mval}}.

###### Measurement Keys {#sec-comid-mkey}

> The types defined for identifying a measured element are uint, or a textual string.

Reviewer: ```suggestion
The types defined for identifying a measured element are uint, or a textual string.
The negative integers range is reserved and MUST NOT be used as a measurement key.
```

~~~ cddl
{::include cddl/measured-element-type-choice.cddl}
~~~

A measured element serves as a local identifier within the scope of the Environment.

> An example of a measured element could be `"PCR0"` from the TCG-specified Trusted Platform Module's bank of platform configuration registers.

Reviewer 2: Maybe we should an example of "element" higher (closer to line 771)?

reconfiguration regions, etc. Measurements comprise raw values, digests, or
status information.

> An environment has one or more measured elements.

Reviewer: I ❤️ "element"

> An environment has one or more measured elements.

Reviewer 3: ```suggestion
An environment has one or more measured elements.
Measured elements are distinguished from each other by the `measurement-map` key.
```

A measured element can have a dedicated measurement, such as a security version number.
Multiple elements could be combined into a single measurement, such as a running hash of multiple files.
A measured element can also have several measurement values, such as a name, version, and digest.
Measurements can have class, instance or group scope.
A `measurement-map`'s scope is determined by the `environment-map` it is paired with.

Class measurements apply generally to all the Attesters in the given class.
Instance measurements apply to a specific Attester instance.  Environments
identified by a class identifier have measurements that are common to the
class. Environments identified by an instance identifier have measurements that
are specific to that instance.

> The supply chain entity that is defined as the authority responsible for providing the the measurements (i.e. Reference Values or Endorsed Values)

Reviewer: ```suggestion
The supply chain entity that is defined as the authority responsible for providing the measurements (i.e. Reference Values or Endorsed Values)
```

is by default the CoRIM signer.

> If different authorities may provide measurement values, the `"__authorities"` statement can be supplied in the `measurement-map`.

Reviewer 3: I think we need some CBOR tweaks here... probably best discussed in the meeting.

Author: As discussed in the meeting, we'll move discussion to issue#244 and resolve the authorities encoding in a follow-up PR.

If the `"__authorities"` statement is used within a condition, at least one of the listed authorities must match the respective ACS entry for the statement to match.

~~~ cddl
{::include cddl/measurement-map.cddl}
~~~

> The following describes each member of the `measurement-map`:

Reviewer 3: ```suggestion
The `measurement-map` contains one or more measured-entities, each is identified by the map key.
It may also contain special values which apply to all measured-entities.
```

Reviewer 2: ```suggestion
The `measurement-map` contains one or more measured elements, each is identified by the map key.
```

> It may also contain special values which apply to all measured-entities.

Should this be the negative value space?
See "fixme-authorities" for context

Reviewer: I agree with Henk that L802 is insufficient.  It needs some clear description of the negative codepoint space which is reserved for common attributes.

> * `"__authorities"`: The cryptographic identities of individuals or organizations that are

Reviewer 3: Suggest moving this to the end and replacing it with something like:
```suggestion
The special map key NULL indicates that the corresponding value is a list of authorities for the measurement.
In an ACS entry this indicates which entity deckared the validity of the measurement.
In a reference value entry this provides a list of the authorities whose authority declaration is required.
```

Author: Oh interesting, I didn't think it was a list of required authorities but rather a list of allowed authorities. At least one must match. Perhaps authorities should be an explicit pattern that is consistent with that interpretation:

authority = $crypto-key-type-choice
authorities = [ + authority ] / #6.562([ + authorities ]) / #6.563([ + authorities ])

We ensure #6.563 has a subset relationship with the list it's matched against, so all listed authority patterns in #6.563 must match some entry of acs-authorities.

This is different from #6.562 since that is any pattern fully matches the value (not an entry of the value, presumed array)

 designated authorities for this measurement. For example, producer of the measurement or a delegated supplier.

* `$measured-element-type-choice`: A map key distinct from `"__authorities"` that names a measured element. See {{sec-comid-mkey}}.

> * `measurement-values-map`: A map value that associates measurement values with the named measured element within the environment. Described in {{sec-comid-mval}}.

Reviewer 3: ```suggestion
When the map value is a uint or string, the map value holds the corresponding measurement values.
Described in {{sec-comid-mval}}.
```

###### Measurement Keys {#sec-comid-mkey}

> The types defined for identifying a measured element are uint, or a textual string different from `"__authorities"`.

Reviewer 3: ```suggestion
The types defined for identifying a measured element are uint, or a textual string.
The meaning of a measurement key is evaluated within the context of its environment-map.
```

~~~ cddl
{::include cddl/measured-element-type-choice.cddl}
~~~

> A measured element serves is a local identifier within the scope of the Environment.

Reviewer: ```suggestion
A measured element serves as a local identifier within the scope of the Environment.
```

> A measured element MUST NOT be used as a refinement on the `environment-map`, since it names an element from a specific Environment's collection of measurements in one appraisal context.

Reviewer 3: Why is this a bad thing?
When matching a reference value against the ACS, I was expecting the mkey to be used as part of finding measurement elements to be compared. Isn't that how the processing happens?

Author: I conceptualize environment map as a "place" and measurement-map as "properties", so I wouldn't say that a particular property from an attester should be considered more refinement on the place it's describing. If there is environment information, I think it should be in the transformation.

{::include cddl/measurement-map.cddl}
~~~

where `mkeyvalue-pair` is

~~~ cddl
{::include cddl/mkeyvalue-pair.cddl}
~~~

The following describes each member of the `measurement-map`:

> * `mkey` (index 0): DEPRECATED (remove for final RFC) An optional identifier of the measured element.

Reviewer 4: I would clarify this in a better way....

I guess you intend to remove the index 0 for `mkey` but want to retain the `mkey` for your mkeyvalue-pair

If none of the series conditions match an ACS Entry, the triple is not matched,
and no Endorsed values are accepted.

> The `"__authorities"` value in `measurement-map` in the stateful environment, if present,

Reviewer 3: Can we define a type, then have the text refer to the type rather than the key which identifies the type?

The ACS is initialized by copying the internal representation of Evidence claims to the ACS.
See {{sec-add-to-acs}}.

> #### The "__authorities" field in Appraisal Claims Set {#sec-authorities}

Reviewer 3: ditto

> 

Reviewer 3: I think the paragraphs above need to be enhanced to say how measurement-values-maps are selected for comparison. Example text for this might be:

```suggestion
The Verifier initializes a temporary "candidate entries" variable with all entries in the ACS where the stateful environment `environment-map` and measured element identifier is a subset of the ACS `environment-map` and measured element identifier.

A stateful environment `environment-map` is a subset of an ACS entry `environment-map` if each field (for example `class`, `instance` etc.) which is present in the stateful environment `environment-map` is also present in the ACS entry, and the CBOR encoded field values in the stateful environment and ACS entry are binary identical.
If a field is not present in the stateful environment `environment-map` then the presence of, and value of, the corresponding ACS entry field does not affect whether the `environment-map`s are subsets.
A stateful environment measured element identifier is a subset of an ACS entry measured element identifier if their CBOR encodings are binary identical.
```

Each entry within `state-triples` uses the syntax of `endorsed-triple-record`.
When an `endorsed-triple-record` appears within `state-triples` it
indicates that the authorities named by `measurement-map`/`"__authorities"`
assert that the actual state of one or more Claims named by the `$measured-element-type-choice` and `measurement-values-map` keys within the
Target Environment, as identified by `environment-map`, have the
measurement values in `measurement-map`[`$measured-element-type-choice`].

ECT authority is represented by cryptographic keys. Authority
is asserted by digitally signing a Claim using the key. Hence, Claims are
added to the ACS under the authority of a cryptographic key.

> Each Claim is encoded as an ECT. The `environment-map`, the `mkey`, and the

Reviewer 3: ```suggestion
Each Claim is encoded as an ECT. The `environment-map`, the `measurement-map` key, and the
```
